### DISCUSSION TITLE
Can we catch IllegalArgumentException from AsyncExecutor.beforeExecute?

### DISCUSSION THREAD

**Discussion by y-yu** [slick/slick#2432] [Questions] [answered]
We encounter this exception and our program freeze:

```
Exception in thread "es-payment-api-prod-master-db-pool-4" java.lang.IllegalArgumentException: requirement failed: count cannot be increased
    at scala.Predef$.require(Predef.scala:337)
    at slick.util.ManagedArrayBlockingQueue.$anonfun$increaseInUseCount$1(ManagedArrayBlockingQueue.scala:43)
    at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
    at slick.util.ManagedArrayBlockingQueue.locked(ManagedArrayBlockingQueue.scala:201)
    at slick.util.ManagedArrayBlockingQueue.increaseInUseCount(ManagedArrayBlockingQueue.scala:42)
    at slick.util.AsyncExecutor$$anon$1$$anon$2.beforeExecute(AsyncExecutor.scala:148)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1126)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:829)
```

We try to update the version to 3.4.0-M1 but it wasn't be solved 😢 

We think that

- If `AsyncExecutor.beforeExecute` throws an exception, threads pool is going to be wrong state? 🤔 
- then our program cannot run any threads, the program freezes but doesn't _die_ 😇 

We don't know why this exception was thrown but we want to kill our program when throwing this exception for now. That's the why we want to catch the exception thrown from `AsyncExecutor.beforeExecute`. Do someone have any idea? 🙏 

### Related issues

- https://github.com/slick/slick/issues/2417
- https://github.com/slick/slick/issues/1856

**Comment by nafg** [comment]
This sounds bad. Any way to reliably reproduce it?

**Reply by y-yu** [reply]
Unfortunately we don't know the reproduce way 😇. We think that following things related with this problem:

- Run `db.run` in another `db.run` (nested `db.run`)
- Run `db.run` in concurrent operations like `Future.sequence`

In case #2417, they solved their problem but it seems not to discover why the exception happened 🤔 So we think that there would be a something wrong.

**Comment by nafg** [comment]
Can you share a program that I can run and get the error at all?

**Reply by y-yu** [reply]
Sorry, the code is our business code 😇 . We are trying to find a minimal working example now.

**Comment by nafg** [comment]
Ok no problem. But I hope you realize that without it I have no way to even try to help.

**Comment by y-yu** [comment]
I see. And we want to catch this exception and exit our program. If someone have seen this exception and avoid it, the way is very welcome.

**Comment by nafg** [comment]
If you can't share the code then maybe your business wants to hire someone to look at it?

**Reply by nafg** [reply]
I'm available for that ;)

**Comment by dwickern** [comment]
I'm seeing this in prod also.

The environment uses `minThreads == maxThreads == maxConnections`. All of the queries are plain `sql"..."`. I can replicate the error consistently in my dev environment by lowering the thread count.

It seems related to the streaming APIs. If I swap out the streaming queries (`DatabaseDef#stream`) with a regular fetch `DatabaseDef#run`) then the problem goes away. @y-yu do you use streaming APIs at all?

**Reply by y-yu** [reply]
Thanks for the reply 😸

- Our environment is also `minThreads == maxThreads == maxConnections` as follows
     ```
     minConnections = 5
     maxConnections = 5
     numThreads = 5
     ```
- However we uses Slick `DBIO` type so we don't use streaming 😢 
    - https://github.com/slick/slick/blob/7a539812f3aeb3c4cf42937c9b820ce939c66188/slick/src/main/scala/slick/dbio/package.scala#L6-L10

**Reply by nafg** [reply]
So do you have something I can run to reproduce it?

**Reply by dwickern** [reply]
Alright, I've got a reproduction. Hopefully this works on your machine too.

https://github.com/dwickern/slick-thread-crash

**Comment by nafg** [comment]
It does.

Not work, that is 😉

I'll have to look at it more later.

**Comment by nafg** [comment]
I wonder if this is related to https://github.com/slick/slick/issues/2430

**Comment by nafg** [comment]
Setting maxConnections to 6 makes https://github.com/dwickern/slick-thread-crash work. I wonder if Slick has an off-by-one error?

**Reply by dwickern** [reply]
I think because the 5 parallel queries + the 1 extra from `db.stream` would use at most 6

**Reply by nafg** [reply]
Oh I see

**Comment by nafg** [comment]
@szeiger @kwark @hvesalai anyone that understands the relevant parts of the code available to discuss?

I think this mostly comes down to behavior merged in https://github.com/slick/slick/pull/1461 by @kwark 

For one thing, https://github.com/slick/slick/pull/1461#issuecomment-205755838 says

> or beforeExecute() method will throw because the requirement in increaseInUseCount() will be invalid. This causes the worker to complete abruptly and a new worker will be created, which will be forced to take an item from the queue. During my tests I have not yet seen this beforeExecute throw an exception, but theoretically it should cause no harm

But it sounds like beforeExecute throwing is just resulting in lack of progress?

**Reply by nafg** [reply]
I don't think it's right anyway, based on the source of ThreadPoolExecutor#runWorker.

I submitted https://github.com/slick/slick/pull/2619

**Comment by nafg** [comment]
@y-yu @dwickern can you both test #2619 please

**Comment by nafg** [comment, marked as answer]
@dwickern gave the go ahead so it should be merged soon and then I plan to release 3.5.0-M1
